File: frontend/src/types/attachment.ts
import { MediaType } from './prompt'

// 文件验证规则接口
export interface FileValidationRule {
  maxSize: number // 最大文件大小（字节）
  allowedTypes: string[] // 允许的 MIME 类型
  allowedExtensions: string[] // 允许的文件扩展名
}

// 媒体类型配置接口
export interface MediaTypeConfig {
  type: MediaType
  label: string
  description: string
  icon: string
  validation: FileValidationRule
  supportsThumbnail: boolean
  supportsPreview: boolean
}

// 文件上传配置接口
export interface UploadConfig {
  maxFileSize: number
  maxFiles: number
  allowedTypes: string[]
  chunkSize?: number
  timeout?: number
}

// 缩略图配置接口
export interface ThumbnailConfig {
  width: number
  height: number
  quality: number
  format: 'jpeg' | 'png' | 'webp'
}

// 附件预览配置接口
export interface PreviewConfig {
  enableImagePreview: boolean
  enableVideoPreview: boolean
  enableAudioPreview: boolean
  enableDocumentPreview: boolean
  maxPreviewSize: number
}

// 文件处理选项接口
export interface FileProcessingOptions {
  generateThumbnail: boolean
  extractMetadata: boolean
  validateContent: boolean
  thumbnailConfig?: ThumbnailConfig
}

// 附件查询参数接口
export interface AttachmentQueryParams {
  prompt_id?: number
  mime_type?: string
  file_size_min?: number
  file_size_max?: number
  created_after?: string
  created_before?: string
  limit?: number
  offset?: number
  sort_by?: 'created_at' | 'file_size' | 'filename'
  sort_order?: 'asc' | 'desc'
}

// 附件统计信息接口
export interface AttachmentStats {
  total_count: number
  total_size: number
  by_type: Record<string, {
    count: number
    size: number
  }>
  by_prompt: Record<number, {
    count: number
    size: number
  }>
}

// 存储配置接口
export interface StorageConfig {
  type: 'local' | 's3' | 'oss'
  basePath: string
  baseUrl: string
  maxStorageSize?: number
  cleanupPolicy?: {
    enabled: boolean
    maxAge: number // 天数
    maxSize: number // 字节
  }
}

// 错误类型常量
export const AttachmentError = {
  FILE_TOO_LARGE: 'FILE_TOO_LARGE',
  INVALID_FILE_TYPE: 'INVALID_FILE_TYPE',
  UPLOAD_FAILED: 'UPLOAD_FAILED',
  DOWNLOAD_FAILED: 'DOWNLOAD_FAILED',
  DELETE_FAILED: 'DELETE_FAILED',
  NOT_FOUND: 'NOT_FOUND',
  PERMISSION_DENIED: 'PERMISSION_DENIED',
  STORAGE_FULL: 'STORAGE_FULL',
  NETWORK_ERROR: 'NETWORK_ERROR',
  INVALID_CONTENT: 'INVALID_CONTENT'
} as const

export type AttachmentErrorType = typeof AttachmentError[keyof typeof AttachmentError]

// 附件错误详情接口
export interface AttachmentErrorDetail {
  code: AttachmentErrorType
  message: string
  details?: Record<string, any>
  file?: string
  timestamp: string
}

// 默认媒体类型配置
export const DEFAULT_MEDIA_TYPE_CONFIGS = {
  [MediaType.TEXT]: {
    type: MediaType.TEXT,
    label: '文本',
    description: '纯文本内容',
    icon: 'Document',
    validation: {
      maxSize: 0,
      allowedTypes: [],
      allowedExtensions: []
    },
    supportsThumbnail: false,
    supportsPreview: true
  },
  [MediaType.IMAGE]: {
    type: MediaType.IMAGE,
    label: '图片',
    description: '图片文件（JPEG、PNG、GIF、WebP）',
    icon: 'Picture',
    validation: {
      maxSize: 10 * 1024 * 1024, // 10MB
      allowedTypes: ['image/jpeg', 'image/png', 'image/gif', 'image/webp'],
      allowedExtensions: ['.jpg', '.jpeg', '.png', '.gif', '.webp']
    },
    supportsThumbnail: true,
    supportsPreview: true
  },
  [MediaType.DOCUMENT]: {
    type: MediaType.DOCUMENT,
    label: '文档',
    description: '文档文件（PDF、Word、Excel、PowerPoint）',
    icon: 'Document',
    validation: {
      maxSize: 50 * 1024 * 1024, // 50MB
      allowedTypes: [
        'application/pdf',
        'application/msword',
        'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
        'application/vnd.ms-excel',
        'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
        'application/vnd.ms-powerpoint',
        'application/vnd.openxmlformats-officedocument.presentationml.presentation'
      ],
      allowedExtensions: ['.pdf', '.doc', '.docx', '.xls', '.xlsx', '.ppt', '.pptx']
    },
    supportsThumbnail: true,
    supportsPreview: false
  },
  [MediaType.AUDIO]: {
    type: MediaType.AUDIO,
    label: '音频',
    description: '音频文件（MP3、WAV、AAC、OGG）',
    icon: 'Headphone',
    validation: {
      maxSize: 100 * 1024 * 1024, // 100MB
      allowedTypes: ['audio/mpeg', 'audio/wav', 'audio/aac', 'audio/ogg'],
      allowedExtensions: ['.mp3', '.wav', '.aac', '.ogg']
    },
    supportsThumbnail: false,
    supportsPreview: true
  },
  [MediaType.VIDEO]: {
    type: MediaType.VIDEO,
    label: '视频',
    description: '视频文件（MP4、AVI、MOV、WebM）',
    icon: 'VideoPlay',
    validation: {
      maxSize: 500 * 1024 * 1024, // 500MB
      allowedTypes: ['video/mp4', 'video/avi', 'video/quicktime', 'video/webm'],
      allowedExtensions: ['.mp4', '.avi', '.mov', '.webm']
    },
    supportsThumbnail: true,
    supportsPreview: true
  }
} as const

// 默认上传配置
export const DEFAULT_UPLOAD_CONFIG: UploadConfig = {
  maxFileSize: 100 * 1024 * 1024, // 100MB
  maxFiles: 10,
  allowedTypes: [
    // 图片
    'image/jpeg', 'image/png', 'image/gif', 'image/webp',
    // 文档
    'application/pdf',
    'application/msword',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'application/vnd.ms-excel',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    'application/vnd.ms-powerpoint',
    'application/vnd.openxmlformats-officedocument.presentationml.presentation',
    // 音频
    'audio/mpeg', 'audio/wav', 'audio/aac', 'audio/ogg',
    // 视频
    'video/mp4', 'video/avi', 'video/quicktime', 'video/webm'
  ],
  chunkSize: 1024 * 1024, // 1MB
  timeout: 30000 // 30秒
}

// 默认缩略图配置
export const DEFAULT_THUMBNAIL_CONFIG: ThumbnailConfig = {
  width: 200,
  height: 200,
  quality: 85,
  format: 'jpeg'
}

// 默认预览配置
export const DEFAULT_PREVIEW_CONFIG: PreviewConfig = {
  enableImagePreview: true,
  enableVideoPreview: true,
  enableAudioPreview: true,
  enableDocumentPreview: false,
  maxPreviewSize: 10 * 1024 * 1024 // 10MB
}

File: app/schemas/attachment.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator


class AttachmentBase(BaseModel):
    """附件基础模式"""
    filename: str = Field(..., max_length=255, description="存储文件名")
    original_filename: str = Field(..., max_length=255, description="原始文件名")
    file_size: int = Field(..., ge=0, description="文件大小（字节）")
    mime_type: str = Field(..., max_length=100, description="MIME 类型")

    @field_validator('file_size')
    @classmethod
    def validate_file_size(cls, v: int) -> int:
        """验证文件大小限制"""
        max_size = 10 * 1024 * 1024  # 10MB
        if v > max_size:
            raise ValueError(f"文件大小不能超过 {max_size // (1024 * 1024)}MB")
        return v

    @field_validator('mime_type')
    @classmethod
    def validate_mime_type(cls, v: str) -> str:
        """验证 MIME 类型格式"""
        if not v or '/' not in v:
            raise ValueError("MIME 类型格式无效")
        
        # 支持的 MIME 类型列表
        allowed_types = {
            # 图片类型
            'image/jpeg', 'image/png', 'image/gif', 'image/webp', 'image/bmp',
            # 文档类型
            'application/pdf', 'application/msword', 
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            'application/vnd.ms-excel',
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
            'application/vnd.ms-powerpoint',
            'application/vnd.openxmlformats-officedocument.presentationml.presentation',
            'text/plain', 'text/csv',
            # 音频类型
            'audio/mpeg', 'audio/wav', 'audio/ogg', 'audio/mp4',
            # 视频类型
            'video/mp4', 'video/avi', 'video/mov', 'video/wmv', 'video/webm'
        }
        
        if v not in allowed_types:
            raise ValueError(f"不支持的文件类型: {v}")
        
        return v


class AttachmentCreate(AttachmentBase):
    """附件创建模式"""
    prompt_id: int | None = Field(None, ge=1, description="关联的提示词 ID（可选，用于临时上传）")
    file_path: str = Field(..., max_length=500, description="文件存储路径")
    thumbnail_path: str | None = Field(None, max_length=500, description="缩略图路径")
    file_metadata: dict[str, Any] | None = Field(None, description="文件元数据")


class AttachmentRead(AttachmentBase):
    """附件读取模式"""
    id: int
    prompt_id: int | None = None  # 允许为空
    thumbnail_path: str | None = None
    file_metadata: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime
    download_url: str = Field(..., description="文件下载链接")
    thumbnail_url: str | None = Field(None, description="缩略图链接")

    model_config = ConfigDict(from_attributes=True)


class AttachmentUpdate(BaseModel):
    """附件更新模式"""
    filename: str | None = Field(None, max_length=255, description="存储文件名")
    file_metadata: dict[str, Any] | None = Field(None, description="文件元数据")


class AttachmentListResponse(BaseModel):
    """附件列表响应模式"""
    items: list[AttachmentRead]
    total: int = Field(..., ge=0, description="附件总数")


class AttachmentUploadResponse(BaseModel):
    """文件上传响应模式"""
    attachment: AttachmentRead
    message: str = Field(default="文件上传成功", description="响应消息")


__all__ = [
    "AttachmentBase",
    "AttachmentCreate", 
    "AttachmentRead",
    "AttachmentUpdate",
    "AttachmentListResponse",
    "AttachmentUploadResponse"
]